File: examples/tlplay-python/StatusBar.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

class Widget(ftk.IWidget):

    def __init__(self, context, app, actions, parent = None):
        ftk.IWidget.__init__(self, context, "StatusBar.Widget", parent)

        self._logLabel = ftk.Label(context)
        self._logLabel.hStretch = ftk.Stretch.Expanding
        self._logLabel.marginRole = ftk.SizeRole.MarginInside

        self._infoLabel = ftk.Label(context)
        self._infoLabel.marginRole = ftk.SizeRole.MarginInside

        self._layout = ftk.HorizontalLayout(context, self)
        self._layout.spacingRole = ftk.SizeRole.SpacingSmall
        self._logLabel.parent = self._layout
        ftk.Divider(context, ftk.Orientation.Horizontal, self._layout)
        self._infoLabel.parent = self._layout

        self._logObserver = ftk.ListObserverLogItem(
            context.getSystemByName("ftk::LogSystem").observeLogItems,
            self._logUpdate)

        self._playerObserver = tl.timeline.ValueObserverPlayer(
            app.getDocumentModel().observePlayer(),
            self._infoUpdate)

        self._logTimer = ftk.Timer(context)

    def setGeometry(self, value):
        ftk.IWidget.setGeometry(self, value)
        self._layout.setGeometry(value)
    
    def sizeHintEvent(self, event):
        self.setSizeHint(self._layout.sizeHint)

    def _logUpdate(self, logItems):
        if logItems:
            text = ""
            for item in logItems:
                pieces = []
                if ftk.LogType.Error == item.type:
                    pieces = item.toString().split('\n')
                if pieces:
                    text = pieces[0]
            self._logLabel.text = text
            if text:
                self._logTimer.start(5.0, self._clearLog)

    def _clearLog(self):
        self._logLabel.text = ""

    def _infoUpdate(self, player):
        text = []
        tooltip = []
        if player:
            path = player.path
            text.append(path.fileName)
            tooltip.append(path.get())

            ioInfo = player.ioInfo
            if ioInfo.video:
                videoInfo = ioInfo.video[0]
                s = "video: {}x{}:{:.2f} {}".format(
                    videoInfo.size.w,
                    videoInfo.size.h,
                    videoInfo.aspect,
                    videoInfo.type)
                text.append(s)
                tooltip.append(s)

            if ioInfo.audio.isValid:
                s = "audio: {} {} {}".format(
                    ioInfo.audio.channelCount,
                    ioInfo.audio.dataType,
                    ioInfo.audio.sampleRate)
                text.append(s)
                tooltip.append(s)

        self._infoLabel.text = ", ".join(text)
        self._infoLabel.tooltip = "\n".join(tooltip)

File: lib/tlRender/PlayApp/StatusBar.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the tlRender project.

#include "StatusBar.h"

#include "App.h"
#include "FilesModel.h"

#include <ftk/UI/Divider.h>
#include <ftk/Core/Format.h>

namespace tl
{
    namespace play
    {
        void StatusBar::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            IWidget::_init(context, "StatusBar", parent);

            _layout = ftk::HorizontalLayout::create(context, shared_from_this());
            _layout->setSpacingRole(ftk::SizeRole::SpacingSmall);

            _labels["Log"] = ftk::Label::create(context, _layout);
            _labels["Log"]->setHStretch(ftk::Stretch::Expanding);
            _labels["Log"]->setMarginRole(ftk::SizeRole::MarginInside);

            ftk::Divider::create(context, ftk::Orientation::Horizontal, _layout);
            _labels["Info"] = ftk::Label::create(context, _layout);
            _labels["Info"]->setMarginRole(ftk::SizeRole::MarginInside);
            
            _logTimer = ftk::Timer::create(context);

            _logObserver = ftk::ListObserver<ftk::LogItem>::create(
                context->getLogSystem()->observeLogItems(),
                [this](const std::vector<ftk::LogItem>& value)
                {
                    _logUpdate(value);
                });

            _playerObserver = ftk::Observer<std::shared_ptr<Player> >::create(
                app->getFilesModel()->observePlayer(),
                [this](const std::shared_ptr<Player>& value)
                {
                    _infoUpdate(value);
                });
        }

        StatusBar::~StatusBar()
        {}

        std::shared_ptr<StatusBar> StatusBar::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<IWidget>& parent)
        {
            auto out = std::shared_ptr<StatusBar>(new StatusBar);
            out->_init(context, app, parent);
            return out;
        }

        ftk::Size2I StatusBar::getSizeHint() const
        {
            return _layout->getSizeHint();
        }

        void StatusBar::setGeometry(const ftk::Box2I& value)
        {
            IWidget::setGeometry(value);
            _layout->setGeometry(value);
        }

        void StatusBar::_logUpdate(const std::vector<ftk::LogItem>& value)
        {
            for (const auto& i : value)
            {
                switch (i.type)
                {
                case ftk::LogType::Error:
                {
                    const std::string text = ftk::getLabel(i, true);
                    _labels["Log"]->setText(text);
                    _labels["Log"]->setTooltip(text);
                    _logTimer->start(
                        std::chrono::seconds(5),
                        [this]
                        {
                            _labels["Log"]->setText(std::string());
                            _labels["Log"]->setTooltip(std::string());
                        });
                    break;
                }
                default: break;
                }
            }
        }

        void StatusBar::_infoUpdate(const std::shared_ptr<Player>& player)
        {
            std::vector<std::string> text;
            std::vector<std::string> tooltip;
            if (player)
            {
                const auto& path = player->getPath();
                text.push_back(path.getFileName());
                tooltip.push_back(path.get());

                const auto& ioInfo = player->getIOInfo();
                if (!ioInfo.video.empty())
                {
                    const auto& videoInfo = ioInfo.video.front();
                    const std::string s = ftk::Format("video: {0}x{1}:{2} {3}").
                        arg(videoInfo.size.w).
                        arg(videoInfo.size.h).
                        arg(videoInfo.getAspect(), 2).
                        arg(videoInfo.type);
                    text.push_back(s);
                    tooltip.push_back(s);
                }

                if (ioInfo.audio.isValid())
                {
                    const std::string s = ftk::Format("audio: {0}").
                        arg(getLabel(ioInfo.audio));
                    text.push_back(s);
                    tooltip.push_back(s);
                }
            }
            _labels["Info"]->setText(ftk::join(text, ", "));
            _labels["Info"]->setTooltip(ftk::join(tooltip, "\n"));
        }
    }
}
